src/cli.js

```diff
@@ -17,11 +17,28 @@ const SCHEMASTORE_CATALOG_URL =
   "https://www.schemastore.org/api/json/catalog.json";
 const CACHE_DIR = path.join(os.tmpdir(), "flat-cache");
 
-async function getSchemaUrlForFilename(filename, cache, ttl) {
-  const { schemas } = await cachedFetch(SCHEMASTORE_CATALOG_URL, cache, ttl);
+async function getSchemaUrlForFilename(filename, cache, ttl, catalogs) {
+  let schemaLocation;
+  try {
+    const schemas = catalogs.flatMap((c) =>
+      JSON.parse(fs.readFileSync(c, "utf8").toString()).schemas.map((s) => {
+        if (!isUrl(s.url)) {
+          s.url = path.resolve(path.dirname(c), s.url);
+        }
+        return s;
+      })
+    );
+    schemaLocation = getSchemaMatchForFilename(schemas, filename);
+  } catch {
+    const { schemas } = await cachedFetch(SCHEMASTORE_CATALOG_URL, cache, ttl);
+    schemaLocation = getSchemaMatchForFilename(schemas, filename);
+  }
+  return schemaLocation;
+}
```

Comment:
There are a few problems I can see with this approach.

- I think `--catalogs` should probably accept both a url or a local file (in the same way that `--schema` does). Custom remote catalogs should also be cached
- I don't think we should attempt to merge the catalogs together here. Lets try and match the file in the first schema catalog first, then fall back to the second one, then the third one, etc.. (so there is an order of priority)
- This code (with the try/catch) behaves very strangely (IMO) if we find >1 matches for a given file in a catalog. It treats it the same as zero matches and ploughs on to schemastore. Conflicting `fileMatch` should be a hard error
- Once there are multiple catalogs in play, I think we need to make it clear to the user which catalog the schema we've chosen came from

I think we can probably think of `catalogs` as always being an array which we iterate and implicitly contains `SCHEMASTORE_CATALOG_URL` as the last item in the array (so schemastore is not 'special'). Pulling all that together, I think my suggested implementation of this approach would be something like:

```js
async function getFromUrlOrFile(location, cache, ttl) {
  return isUrl(location)
    ? await cachedFetch(location, cache, ttl)
    : JSON.parse(fs.readFileSync(location, "utf8").toString());
}

async function getSchemaUrlForFilename(catalogs, filename, cache, ttl) {
  for (const [i, catalog] of catalogs.entries()) {
    const { schemas } = await getFromUrlOrFile(catalog, cache, ttl);
    const matches = getSchemaMatchesForFilename(schemas, filename);
    if (matches.length === 1) {
      console.log(`ℹ️ Found schema in ${catalog} ...`);
      return matches[0].url; // Match found. We're done.
    }
    if (matches.length === 0 && i < catalogs.length - 1) {
      continue; // No match found. Try the next catalog in the array.
    }
    if (matches.length > 1) {
      // We found >1 matches in the same catalog. This is always a hard error.
      console.log(
        `Found multiple possible schemas for ${filename}. Possible matches:`
      );
      matches.forEach(function (match) {
        console.log(`${match.description}: ${match.url}`);
      });
    }
    // Either we found >1 matches in the same catalog or we found 0 matches
    // in the last catalog and there are no more catalogs left to try.
    throw new Error(`❌ Could not find a schema to validate ${filename}`);
  }
}

function getSchemaMatchesForFilename(schemas, filename) {
  const matches = [];
  schemas.forEach(function (schema) {
    if ("fileMatch" in schema) {
      if (schema.fileMatch.includes(path.basename(filename))) {
        matches.push(schema);
        return;
      }
      for (const glob of schema.fileMatch) {
        if (minimatch(path.normalize(filename), glob)) {
          matches.push(schema);
          break;
        }
      }
    }
  });
  return matches;
}
```

..and then in `Validator`:

```js
const schemaLocation =
  args.schema ||
  (await getSchemaUrlForFilename(
    [].concat(args.catalogs).concat([SCHEMASTORE_CATALOG_URL]),
    filename,
    cache,
    ttl,
  ));
const schema = await getFromUrlOrFile(schemaLocation, cache, ttl);
```

Reply:
Fixed in dad46a

Reply:
yes you're right - `concat()` doesn't mutate :+1: